### DISCUSSION TITLE
Add support for block-level scope

### DISCUSSION THREAD

**Discussion by mwolfe02** [twinbasic/twinbasic#1463] [Ideas]
For a brief primer on block-level scope (and VBA's current lack of support for it), see here: [Block-Level Scope in VBA...Does Not Exist](https://nolongerset.com/block-level-scope/)

My question is, should we add such support to twinBASIC?

I had a [brief discussion with Wayne on Twitter](https://twitter.com/NoLongerSet/status/1614478361241477120) when someone else asked about twinBASIC support for block-level code.  Wayne shared his concerns:

> I doubt a setting/attribute would cut it, as I agree it would be confusing.  So I expect only explicit new syntax would be considered.   Given how hard it is to gain dev support for new syntax, I would think it is unlikely that there will be movement on this any time soon.

I agree with all points.  After a few days of thinking about it, I realized that the `Private` keyword would be perfect to reuse for this situation.  If you try to declare a local variable as `Private` in VBA you receive the following compile error:

> Invalid attribute in Sub or Function

Thus, we could repurpose that keyword to explicitly enable block-level scope in twinBASIC.  For example:

```
Sub TestBlockScope()
    If True Then
        Private foo As Long
    End If
    Debug.Print foo  '<-- This would raise a compile error such as,
                     '    "'foo' is inaccessible due to its protection level"
End Sub
```

Of course, I have no idea how difficult this would be for *Wayne to implement*. :grin:

Thoughts?

**Comment by WaynePhillipsEA** [comment]
Technically, there would be no major issue to implement such a feature.

Personally, I quite like the re-use of the `Private` keyword, or perhaps even `Private Dim`(?)

**Reply by mwolfe02** [reply]
I prefer `Private MyVar` for brevity and consistency with the existing module-level private variable declaration.  On the other hand, `Private Dim MyVar` being completely unsupported in VBx has advantages in that it's clearly new syntax and would help further distinguish it as a new feature of twinBASIC.  I'm curious what others think.

**Reply by mburns08109** [reply]
Why "Private" over, say "Local" or a variant of Dim like LDim (short for "Local Dim")?

**Comment by saturday220514** [comment]
This might be a stupid suggestion but if 'Private Dim' is possible then what about 'Block Dim' for greater clarity?

**Reply by bclothier** [reply]
FWIW, I'd go with `Local`. This is a illegal syntax in VBx and works for both `Local x` and `Local x As Long`.

**Reply by mwolfe02** [reply]
My thinking is that `Private` is already a keyword so we wouldn't be introducing new keywords into the language.

**Reply by mwolfe02** [reply]
After some quick testing, it looks like `Local` is also a keyword but `Block` is not.

**Reply by WaynePhillipsEA** [reply]
The parser reads left-to-right.  `Private` is already a reserved keyword used elsewhere in the VBx syntax, so in this context the parser knows immediately upon encountering the `Private` symbol that this statement is the new block scoped Dim, since that symbol is invalid in VBx in the same context.

If you don't reuse an existing keyword then things get a lot more complicated and you run the high risk of creating breaking changes to existing syntax.

**Reply by WaynePhillipsEA** [reply]
`Local` is indeed another good candidate.

**Reply by mwolfe02** [reply]
So `Local` isn't listed in this [list of VBA keywords](https://learn.microsoft.com/en-us/office/vba/language/reference/keywords-visual-basic-for-applications), but `Dim Local As Long` is a syntax error.  What is `Local` currently used for in VBx?

**Reply by bclothier** [reply]
Nothing meaningful. It's legal to do `On Local Error` but the compiler treats it as `On Error`. `On Local Error` once meant something in some past incarnation but has no use in the current VBx versions.

**Reply by mwolfe02** [reply]
Found this on [stackoverflow](https://stackoverflow.com/a/29395669/154439):

> On Local Error is a remnant from older versions of BASIC (before Visual Basic for Applications). In current versions it acts the same as On Error and there is no particular reason to use it any more.

So `Local` is basically a currently unused keyword that is only still around for backwards-compatibility reasons.  Seems like a perfect candidate.  I think I prefer `Local` to `Private` as I feel like it's a little closer to the semantic meaning of block-level scope.  Either would be a good option, though.

**Reply by bclothier** [reply]
Curiously, it's not even mentioned in the [MS-VBAL reserved identifier](https://learn.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/7df907cb-ab6c-40d3-aa81-272742ce00c3). Seems odd since the compiler obviously treats it as reserved.

**Reply by mwolfe02** [reply]
Yeah, that was the first place I checked :joy:

**Reply by saturday220514** [reply]
I think I get it.
It had not occurred to me that someone might have declared a procedure named Block.
It's way back in the mists of time but I have a vague recollection of a very small breaking change going from vb4 to vb6 that was fixed 
by my changing the name of something. I presume that when I opened the project in vb6 it complained about some word as
being reserved but I can't remember.

**Reply by bclothier** [reply]
MS-VBAL's [Error Handling Statement](https://learn.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/e2561165-c99a-444b-8bc0-be60a196867a) convinces me that it was an unintentional omission from the specifications since it does not even mention the `Local` even though it's obviously legal syntax. Consequently, it was probably omitted from the reserved identifier in error as well.

**Comment by Greedquest** [comment]
I was thinking about method scope. Public and private to me say you should be able to see that variable _outside_ the function.

```vba
Sub A()
    Public i As Long
End Sub

Sub B()
    Debug.Print A.i 
End Sub
```

I've seen this sort of thing can be used for recursion, or for decorating method with attributes. Not very VBAish though as functions are not first class objects. But sometimes it would be nice to get at the Static variables from inside a function - e.g sharing state between a Property Get/Set pair without module level variables.

 I think Public/Private would be better for in/out procedure scope rather than in/out block scope.

**Reply by mwolfe02** [reply]
Good points.  Further makes the case for `Local` rather than `Private`.

**Comment by Greedquest** [comment]
Syntax aside, are we in favour of block scope functionality? And clear on what it means? 

Personally, I am in favour, I think it is a natural extension of the existing access modification rules. I think the main benefit is that it keeps the visibility (and lifetime?) of a variable as small and protected as possible, making you less likely to accidentally modify it with a typo, particularly in procedures that span more than one screen.

I think it should also include the same "variable used before initialisation" error for consistency. I think something to bear in mind is nested blocks. Python uses `nonlocal` to specify a variable has come from a parent scope - e.g if you declare a function inside a function and the child accesses a variable from the parent. However VBA had no precedent because everything is generally pretty global so I think accessing block scope variables in a parent block should be permitted.

```vba
For i = 1 to 5
    BlockScoped item As Range = Cells(i)
    For j = 1 to 5
        NonLocal item 'do we need this to grab `item` from enclosing scope? I don't think so
        item.Value = item.Value + j
    Next j
Next i
```

**Reply by bclothier** [reply]
I'd want the same scoping rules to apply to keep things simple. Therefore:

```
For i = 1 to 5
    Local item As Range = Cells(i)
    For j = 1 to 5
        Local item 'shadows the previous 'Local item' in outer block scope`
        item.Value = item.Value + j
    Next j
Next i
```

That said, I would definitely want to flag that with hint/warning because of ambiguity.

**Reply by mwolfe02** [reply]
I think we'd want to match [VB.NET's behavior](https://learn.microsoft.com/en-US/dotnet/visual-basic/programming-guide/language-features/declared-elements/scope#block-scope) as closely as possible (with the obvious exception of using `Local` or `Private` in place of `Dim`).

**Comment by wqweto** [comment]
Btw, what is the prior art with VB.Net? Is there block scope implemented? Can variables be (forward) used within local scope before actual `Dim` statement?

**Reply by mwolfe02** [reply]
It appears VB.NET [enforces Block Scope by default](https://learn.microsoft.com/en-US/dotnet/visual-basic/programming-guide/language-features/declared-elements/scope#block-scope).  This note sheds a bit more light on some of the technical details:

> Even if the scope of a variable is limited to a block, its lifetime is still that of the entire procedure. If you enter the block more than once during the procedure, each block variable retains its previous value. To avoid unexpected results in such a case, it is wise to initialize block variables at the beginning of the block.

Like with VBA, variables cannot be used before being declared, even if the declaration occurs within the same scope: [Block-Level Scope Restrictions in VB.NET](https://nolongerset.com/block-level-scope/#block-level-scope-restrictions-in-vbnet)

**Reply by bclothier** [reply]
They use the ~same~ similar scoping rules as C# so we cannot look to VB.NET for inspiration.

```
Imports System
				
Public Module Module1
	Public Sub Main()
		Console.WriteLine("Hello World")
		
		If true then
			dim x as long = 1
			Console.WriteLine(x)			
		end if
		
		If true then
			dim x as long = 2
			Console.WriteLine(x)
		end if
	End Sub
End Module
```
[Test it out using fiddler](https://dotnetfiddle.net/).

**Reply by wqweto** [reply]
> Even if the scope of a variable is limited to a block, its lifetime is still that of the entire procedure.

Wow! Is this like taking the worst of both worlds here :-))

So although declared within a block scope it's actually declared at procedure level but remains hidden for the rest of the procedure blocks. In VB6 at least it's not hidden but as a consequence cannot be shadowed in nested (or not) blocks.

**Reply by Greedquest** [reply]
> Wow! Is this like taking the worst of both worlds here :-))

Agreed. Variable lifetime should be controlled by the block scope. Unless it is Static.

**Comment by FullValueRider** [comment]
We can already emulate Block scoping to an extent by using Dim and a variable name with a number appended.   Personally I quite dislike shadowing variable names due to block scoping and much pefer distinct naming .  In the above example I'd be using x_l2, x_l2m x_l3 etc.

So I propose a firm no to the introduction of block scoping at the compiler level.

**Reply by Greedquest** [reply]
Isn't that an argument against name shadowing specifically rather than block scoping in general? There could be a compiler warning to discourage name shadowing or prohibit it entirely. But still have the other benefits of block scoping.

**Reply by FullValueRider** [reply]
Not really.  In a method I'd only want to see a variable  name declared once.  IMHO using the same variable name in different blocks within a single method is just cognitive overload.  Not wrong, just overload.

The twinBasic enhancement to to allow assignment at the time of dim makes it simpler to add a level based variable name.

**Reply by mwolfe02** [reply]
My preference would be:

- Enforce block scope at the compiler level
- Prohibit name shadowing at the compiler level

```
Sub TestBlockScope()
    If True Then
        Local foo As Long
    End If
    Debug.Print foo  '<-- This would raise a compile error such as,
                     '    "'foo' is inaccessible due to its protection level"

    If True Then
        Local foo As String   '<-- This would raise a compile error such as,
                              '    "duplicate declaration in procedure" 
    End If
End Sub
```

Note that the current compile error for the second case in VBA is "Duplicate declaration in current scope."  That would be an imprecise message with block scoping.

This would provide the benefits of block scoping for developers that are aware of them (e.g., turning runtime errors into compile errors), while being less confusing for longtime VBx developers where block scoping has never been a thing in the past.

**Reply by bclothier** [reply]
My concern is that this reduces the `Local` into just a more restrictive way to declare a variable. IMPOV, the real benefit of block scoping is that we avoid having to think of new names for throwaway variables:
```
If Foo Then
  Local i As Long, x As Long, y As Long
  For i = x To y
    DoFoo
  Next
End if

If Bar Then
  Local i As Long, x As Long, y As Long
  For i = x To y
    DoBar
  Next
End If
```

This sample allows me to reuse the same name and thus avoid having to think/remember what variable to juggle with when I'm in that block. I would feel comfortable using `i` because it is truly a throwaway; all it does is just iterate, and has no additional meaning. 

In current VBx syntax, I have to disambiguate by `FooIndex` and `BarIndex` which are more descriptive but only necessary because I can't use `i` for both. I can't justify using `i1` and `i2` because that usually increase the cognitive load and decrease the readability much more than the verbose `FooIndex`/`BarIndex`.

**Reply by mwolfe02** [reply]
Good counterpoint.  I hadn't considered that use case.

**Reply by Greedquest** [reply]
@bclothier For that usecase I think what we really want is a block scoped variable declared inline

```vba
For i As Long = LBound(args) To UBound(args)
    'i is valid inside the loop only
Next i

For i As Long = 1 To books.Count
    'this i is different to the first one
Next i

For Each b As Book In books
    'book is scoped to inside the loop
Next i
```

But that is syntactic sugar rather than new functionality

**Comment by mburns08109** [comment]
I could favor block scoping, but think it is an unnecessary feature to offer.
What is the true benefit that a block-scoped variable offers over a function-scoped one?

As has been said, cognitive overload is a risk here - I can understand  the need to take such risks, but only when they are truly justified. The VB.Net's lazy-collection scoping rules for this are ridiculous. I'd only support this if the local scope is strictly enforced to the block that defines it. Local variables should be visible to "inner block" scopes as fully read/write and completely fall out of scope (losing any value) at the close of the block.

That would be the only way I can see this even making sense. overloading and magic-tricks to see the overridden outer-block-scoped one while having the same name used in an inner-block is just idiocy and an open invitation to cognitive dissonance (to say the least - I'd actually think I might find axe-murdering Wayne for inflicting this on us as a reasonable solution once my sanity snapped after trying to fully grok this and find a good use-case for it if we were to add that block-scope-overriding-but-still-referencing-the-outer-local-in-the-inner-block-with-an-overridden-same-variable-name crud - screw all that and make it a hard compile error to re-local-Dim the same variable with that same local-dim'd variable already in scope! - both to save all our sanity - and Wayne's life!  ;-) ).

Now, in the case where recursion is in play, such as:
```
Function Recurse(ByVal A as whatever) as whatever
dim Y as whatever

for Y = 1 to somevaluehere
    Local D as somethingorother = A
    If SomethignAboutY = True then Recurse(D)   'don't worry about datatype consistency for this hypothetical example, ok?
Next

end function
```

In that example, D being passed by value presents no issue with scoping.
If D were ByRef, though, we might need to consider the risks there a bit more fully.

Perhaps it would even make sense to require ByVal-only when passing a Local-Dim'd variable out of the block it's defined within?  So it'd have to be Recurse(ByVal D) - in ALL cases?   ...just a thought.

**Reply by Greedquest** [reply]
@mburns use triple backticks ` ```vba ` above and below the code, and no indents needed
````
```vba
'code here
```
````

**Reply by Greedquest** [reply]
```vba
Dim total As Long
Dim i As Long
For i = 1 To 5
    Dim j As Long
    For j = 1 To 10
        Dim k As Long
        For k = 1 To 10
            Arr(i, j, k) = Arr(5 - i, 10 - j, 10 - k) * 2
            total += Arr(i, j, k)
        Next k
    Next j
    If k Mod 2 = 0 Then total *= 2 'oops I meant j not k
Next i
```

Contrived example, `k` only has logical meaning inside that inner loop. However in VBx we can use it outside the loop. Makes the probability of a typo just that little bit higher. Or you can imagine using a variable in a with block to manage lifetime

```vba
With New Excel.Application
    Dim wb As Workbook = .Open("foo.xlsm")
    wb.Range("A1").Value = "Hello"
    .Quit
End With
'...
Debug.Print "Processed workbook: "; wb.Name 'runtime error. But this could be a compiler error with local scoping, easier to track down
```

**Reply by mburns08109** [reply]
Thanks Greedquest.
```
debug.print "Got it!"
```

...and yeah, ok THAT is a good reason for adding this feature - but without this level of explanation, it's really subtle. As a part of documenting this feature, this turning runtime errors into compile-time errors needs to be put forward as an important rationale for having and using this new capability.

**Reply by Greedquest** [reply]
No problem :) although even better:

```vba
Debug.Print "with colour - ```vba don't forget to specify the language"
```

Yes agreed the motivation was not clear in the original post which is why I asked whether block scoping was something people wanted. I think yes.

**Reply by mburns08109** [reply]
@Greedquest,

OK the Block variable is an ActiveX Server object reference, like Excel.Application.

In a circumstance like that I can see where passing an inner block object reference variable out recursively in a ByRef parameter...Yeah, I can definitely see trouble there. Where is that pointer stored - Heap or Stack? (or both? ...due to the recursion's use of stack frames?)

What if the called code tries to manage the COM Server's lifespan?  Excel.Quit, anyone (naturally this would occur in the 97th-level iteration of the recursion loop)? Good luck debugging that.

**Reply by Greedquest** [reply]
@mburns08109 Not sure where pointers are stored, I would be surprised if pointer dereferencing was performance critical in this scenario. You would only need to pass the object ByVal as it is still an object and ByVal/ByRef point to the same object instance. So no semantics changed.

The idea with `Excel.Quit` and a with block is better handled by a _Context Manager_

```vba
Public Class ExcelContext
	Public ReadOnly app As New Excel.Application
	Private Sub Class_Terminate()
		app.Quit
	End Sub
End Class

'...
With New ExcelContext
	On Error Goto SafeExit 'ensure terminate code can run
        Dim wb As Workbook = .app.Open("foo.xlsm")
        wb.Range("A1").Value = "Hello"
    SafeExit:
End With '.Quit is called here
'...
Debug.Print "Processed workbook: "; wb.Name 'runtime error. But this could be a compiler error with local scoping, easier to track down
```

This means `Quit` only happens at the final `End With`, not 97 levels down.

**Reply by mburns08109** [reply]
Oh yeah, I know that, but I was providing you a contrived example in return to illustrate the problems this could lead to.
